src/Bedrock.Framework/Protocols/Protocol.cs

```diff
@@ -87,89 +89,89 @@ public ProtocolReader(ConnectionContext connection, IProtocolReader<TReadMessage
 
         public ConnectionContext Connection { get; }
 
-        public async ValueTask<TReadMessage> ReadAsync(CancellationToken cancellationToken = default)
+        public async ValueTask<ReadResult<TReadMessage>> ReadAsync(CancellationToken cancellationToken = default)
```

Comment:
- Nested structs aren't great but there's nothing we can do about that.
- Should this return the previously read value if Advance isn't called?